DCS: Feature/limit dcs scans to one at a time

adam1985d opened the pull request:
Added `FINISH` state and moved the scan-cleanups to the `FINISH` state.

Added to the DCS class functions and a static data member that adds the capability to limit to a single scan at a time.
Added PRE_SCAN state in which the limitation of single-scan-at-a-time is enforced. When single-scan or continuous-scan conditions are met, the state is changed to PRE_SCAN and only when the no-scan-in-progress condition is met, the state is changed to `TRIGGER_SCAN` which triggers the scan.

The limit to a single scan at a time is independent of the number of DCS tasks created.

The tasks release the ownership of the scan in the new `FINISH` state as part of the scan-cleanup.

In the case that the DCS task performs a retry of the scan, the task will require to take ownership once again - this is to prevent starvation of other tasks.

Fixes: #1094

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.h:
```diff
@@ -90,6 +91,8 @@ class dynamic_channel_selection_task : public task, public std::enable_shared_fr
     bool m_is_single_scan_pending = false;
     bool m_is_single_scan         = false;
+
+    static int m_scanning;
```
```suggestion
    static int m_scanning_task_id = -1; // Currently scanning task id. -1 means no scan in progress.
```

adam1985d replied:
static initialization is done inside the cpp file. added the comment as suggested

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.h:
```diff
@@ -20,6 +20,7 @@
 // Helper MACROs for Enum/String generation
 #define GENERATE_ENUM(ENUM) ENUM,
 #define GENERATE_STRING(STRING) #STRING,
+#define NO_SCAN_IN_PROGRESS -1
```
If you change `m_scanning` to `m_scanning_task_id`, no need for this define.

adam1985d replied:
changed `m_scanning` to `m_scanning_task_id`

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -206,6 +224,13 @@ void dynamic_channel_selection_task::work()
+            TASK_LOG(ERROR) << "scan abort event received for a task that isn't the scanning task";
+        } else {
+            // scan by current task is aborted
+            m_scanning = NO_SCAN_IN_PROGRESS;
```
```suggestion
            m_scanning_task_id = -1;
```

adam1985d replied:
done

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -206,6 +224,13 @@ void dynamic_channel_selection_task::work()
     case eState::ABORT_SCAN: {
         LOG(ERROR) << "aborting scan for mac=" << m_radio_mac << ", last_scan_timestamp is not set";
 
+        if (m_scanning != id) {
```
```suggestion
        if (m_scanning_task_id != id) {
```

adam1985d replied:
done

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -187,6 +195,16 @@ void dynamic_channel_selection_task::work()
+        }
+
+        // scan by current task is complete
+        m_scanning = NO_SCAN_IN_PROGRESS;
```
```suggestion
        m_scanning_task_id = -1; //mark no task is currently scanning
```

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -118,6 +120,12 @@ void dynamic_channel_selection_task::work()
+            break;
+        }
+        // set scanning task id to current task id
+        m_scanning = id;
```
```suggestion
        m_scanning_task_id = id;
```

adam1985d replied:
done

tomereli commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -18,6 +18,8 @@
 std::string s_ar_states[] = {FOREACH_DCS_STATE(GENERATE_STRING)};
 std::string s_ar_events[] = {FOREACH_DCS_EVENT(GENERATE_STRING)};
 
+int dynamic_channel_selection_task::m_scanning = NO_SCAN_IN_PROGRESS;
```
```suggestion
int dynamic_channel_selection_task::m_scanning = -1; //no scan in progress
```

adam1985d replied:
done

tomereli reviewed (approved):
Small changes which I think you should do, but nothing that requires re-review.

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.h:
```diff
@@ -90,6 +91,8 @@ class dynamic_channel_selection_task : public task, public std::enable_shared_fr
     bool m_is_single_scan_pending = false;
     bool m_is_single_scan         = false;
+
+    static int m_scanning;
```
```suggestion
    static int m_scanning_task_id;
```

adam1985d replied:
done

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.h:
```diff
@@ -20,6 +20,7 @@
 // Helper MACROs for Enum/String generation
 #define GENERATE_ENUM(ENUM) ENUM,
 #define GENERATE_STRING(STRING) #STRING,
+#define NO_SCAN_IN_PROGRESS -1
```
No need for this define, we use `-1` for invalid task id.
Maybe should add an `INVALID_TASK_ID` to the base class (but it's not part of this PR anyway).

adam1985d replied:
removed the define

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -118,6 +120,12 @@ void dynamic_channel_selection_task::work()
         break;
     }
     case eState::TRIGGER_SCAN: {
+        if (m_scanning != NO_SCAN_IN_PROGRESS) {
```
When moving to the `TRIGGER_SCAN` state from the `IDLE` state, we also initialize the `m_last_scan_try_timestamp` member. If now the task will block until the previous scan is complete, what impact does it have on this timeout?
Maybe it's better to remain in `IDLE` state? or even in `INIT`?

adam1985d replied:
good point, but I think it is better to move the `m_last_scan_try_timestamp` set point to where we actually trigger the scan - meaning the `TRIGGER_SCAN` state.
will move the set of `m_last_scan_try_timestamp` to `TRIGGER_SCAN` state - right after setting the `m_scanning_task_id`

adam1985d replied:
please re-review. provided a solution that is based on your suggestion.

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -108,23 +108,22 @@ void dynamic_channel_selection_task::work()
-            auto now = std::chrono::steady_clock::now();
-
-            if (now > m_next_scan_timestamp_interval) {
+            if (std::chrono::steady_clock::now() > m_next_scan_timestamp_interval) {
```
I still think that there's no sense to move into the `TRIGGER_SCAN` state if you can't really trigger a scan.
The task should remain in `IDLE` until it's able to trigger a scan.

adam1985d replied:
please see the last fixup - it addresses both.
I will edit the commit message accordingly as part of the squash

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -195,15 +194,15 @@ void dynamic_channel_selection_task::work()
+        if (m_scanning_task_id != id) {
             TASK_LOG(ERROR) << "scan done event received for a task that isn't the scanning task";
             m_last_scan_error_code = beerocks::eChannelScanErrCode::CHANNEL_SCAN_INTERNAL_FAILURE;
             fsm_move_state(eState::ABORT_SCAN);
```
Why do we `ABORT` the scan in this case? How can this even happen?

adam1985d replied:
it shouldn't happen - but wanted to be extra safe here and not update the `m_scanning_task_id` and the results if for some reason this state is reached even though we don't own the scanning-permission

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -457,4 +447,25 @@ void dynamic_channel_selection_task::fsm_move_state(eState new_state)
+    // scan by current task is aborted
+    m_scanning_task_id = -1;
+    return true;
+}
```
Missing new line

adam1985d replied:
true - I intend to run the clang after the squash

vitalybu replied:
Still missing

adam1985d replied:
fixed

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -194,16 +190,13 @@ void dynamic_channel_selection_task::work()
         LOG(TRACE) << "SCAN_DONE";
 
-        if (m_scanning_task_id != id) {
+        if (!finish_scan()) {
```
A better solution may be to add a new `FINISH` state, where you finalize everything.
Then you won't have to do this from multiple states.

adam1985d replied:
I agree - but this is more of refactoring the scan_done and abort_scan cleanups to be in a new state - and less of the actual add of protection.

I added this as part of the PR.

kantera800 replied:
good idea, steering task also use such a state

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -102,28 +102,24 @@ void dynamic_channel_selection_task::work()
     }
     case eState::IDLE: {
-        if (m_is_single_scan_pending) {
+        if (m_is_single_scan_pending && start_scan()) {
```
IMO It's better to call this function right before moving to the `TRIGGER_SCAN` state.
Something like this:
```c++
auto now = std::chrono::steady_clock::now();
if (m_is_single_scan_pending) {
    ...
} else if (database.get_channel_scan_is_enabled(m_radio_mac) {
    ...
} else {
    break;
}

if (start_scan()) {
    m_last_scan_try_timestamp = now;
    fsm_move_state(eState::TRIGGER_SCAN);
}
```

In this case you actually start the scan (`start_scan()`) only when all conditions are met and no other task is currently scanning.

adam1985d replied:
but the solution you suggest updates the `m_last_scan_try_timestamp` for a single scan as well.
also, you don't want to set the data members only to then break due to start_scan() returning false

vitalybu replied:
Also true, but can be solved by adding another state :)
`IDLE` -> `CAN_SCAN` -> `TRIGGER_SCAN` -> ... -> `FINISH`

The FSM will reach the `CAN_SCAN` state only when it's ready to scan.
It'll wait in this state until it can actually trigger a scan, and only then move to `TRIGGER_SCAN`.
`FINISH` will always do all the cleanups (including removing the id of the scanning task).

adam1985d replied:
it is more of `SHOULD_SCAN` rather than `CAN_SCAN` since the `start_scan()` itself is the condition of can-scan

adam1985d replied:
I added a PRE_SCAN state and FINISH state - please re-review

kantera800 commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -432,4 +449,25 @@ void dynamic_channel_selection_task::fsm_move_state(eState new_state)
+        return false;
+    }
+    // scan by current task is aborted
+    m_scanning_task_id = -1;
```
I'm pretty sure that a task can be erased without going through finish_scan().
I'm wondering if this is the only place for clearing this flag.

maybe we should clear it in the destructor
` ~dynamic_channel_selection_task() { m_scanning_task_id = -1 }`

or 

```
void dynamic_channel_selection_task::handle_task_end()
{
  m_scanning_task_id = -1;
}
```

vitalybu replied:
That's a good point, but instead of setting `m_scanning_task_id = -1;`, you should just call `finish_scan();` from the destructor (as it also make checks that the current task if the owner of this flag).

adam1985d replied:
added a fixup commit that calls `finish_scan();` as part of the DTOR

kantera800 commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -103,19 +106,28 @@ void dynamic_channel_selection_task::work()
         if (m_is_single_scan_pending) {
             m_is_single_scan         = true;
             m_is_single_scan_pending = false;
```
(nitpick ) I know its not part of this PR but `m_is_single_scan_pending` and `m_is_single_scan_pending` confused me .
i prefer `m_is_manual_scan` and also would add a comment in code why do we need to treat manual scan differently from continues one .

adam1985d replied:
will add it as a separate hotfix, but for discussion, we use the "single" terminology across the entire DCS feature implementation - why should the task name it differently?
we can change it to `m_handling_single_scan_request` if it would be more clear.

secondly - the `m_is_single_scan_pending` is a flag that indicates that a single scan request was made and haven't been treated yet.
the `m_is_single_scan` is used through the entire task to determine which scan params to read and which scan results to update. where do you think a comment regarding this should be added?

kantera800 replied:
> will add it as a separate hotfix, but for discussion, we use the "single" terminology across the entire DCS feature implementation - why should the task name it differently?
> we can change it to `m_handling_single_scan_request` if it would be more clear.
>
In that case just adding "request" would do 
`m_single_scan_request` 
`m_single_scan_request_pending`

> 
> secondly - the `m_is_single_scan_pending` is a flag that indicates that a single scan request was made and haven't been treated yet.
> the `m_is_single_scan` is used through the entire task to determine which scan params to read and which scan results to update. where do you think a comment regarding this should be added?

i would add it to the header file

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -432,4 +449,25 @@ void dynamic_channel_selection_task::fsm_move_state(eState new_state)
+
+bool dynamic_channel_selection_task::finish_scan()
+{
+    if (m_scanning_task_id != id) {
```
Since now you're also calling this method from the destructor, you should also check that `m_scanning_task_id != -1`, otherwise it will always be printed:
```suggestion
    if (m_scanning_task_id != -1 && m_scanning_task_id != id) {
```

adam1985d replied:
I will add a separate check:
```
if (m_scanning_task_id == -1) {
    return true;
}
```

vitalybu commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -432,4 +449,25 @@ void dynamic_channel_selection_task::fsm_move_state(eState new_state)
+        TASK_LOG(ERROR) << "finish scan is called for a task that isn't the scanning task";
+        return false;
+    }
+    // scan by current task is aborted
```
```suggestion
    
    // scan by current task is finished
```

adam1985d replied:
fixed

adam1985d committed:
controller: dcs-task: log print if any scan condition is met

Added log print if the single-scan condition is met.
Moved "intreval_condition is met" log print before fsm_move_state()
for better readability since the fsm_move_state also prints the state
change.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: dcs-task: move scan cleanup to finish state

added FINISH state to the task.
moved the task cleanups to the finish state.
the SCAN_DONE and ABORT_SCAN change-state to the FINISH state after
performing the state-specific operations.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: dcs-task: infrastructure to limit to one scan only

Preparative Commit.
added a static data member to limit to a single scan at a time.
added start_scan() private function that sets scanning-task-id to
current-task-id and returns true if a scan is not in progress,
otherwise false.
added finish_scan() private function that sets scanning-task-id to -1,
which represents no-scan-is-in-progress and returns true. if the
current-task-id is not equal to the current-task-id, the function
returns false.

added PRE_SCAN state, that the start_scan() check will be performed at
before moving to TRIGGER_SCAN state.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: dcs-task: limit to one scan at a time

changed IDLE state to move to PRE_SCAN state when a single-scan or
continuous-scan requirement is met.
added start_scan() check, in PRE_SCAN state, if successful, change-state
to TRGGER_SCAN state.

updated m_last_scan_try_timestamp in PRE_SCAN state instead of IDLE,
when start_scan() is successful.

Signed-off-by: Adam Dov <adamx.dov@intel.com>
